File: C++/level2/튜플.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

bool compare(vector<int> first, vector<int> second){
    if(first.size() < second.size())
        return true;
    return false;
}

vector<int> solution(string s) {
    vector<int> answer;
    vector<vector<int>> cache;
    int num = 0;
    vector<int> temp;
    vector<int> empty;

    for(int i=1; i<s.size()-1; i++){
        if(num != 0 && s[i] >= '0' && s[i] <= '9'){
            num *= 10;
            num += (s[i] - '0');
        }
        else if(s[i] == '{'){
            i++;
            num = (s[i] - '0');
        }
        else if(s[i] == '}'){
            temp.push_back(num);
            cache.push_back(temp);
            temp = empty;
            num = 0;
            i++;
        }
        else if(s[i] == ','){
            temp.push_back(num);
            num = 0;
        }
        else
            num = (s[i] - '0');
    }


    sort(cache.begin(), cache.end(), compare);

    for(int i=0; i<cache.size(); i++)
        sort(cache[i].begin(), cache[i].end());

    answer.push_back(cache[0][0]);

    for(int i=0; i<cache.size()-1; i++){
        for(int j=0; j<cache[i].size(); j++){
            if(cache[i][j] != cache[i+1][j]){
                answer.push_back(cache[i+1][j]);
                break;
            }
            if(j == cache[i].size()-1 && cache[i][j] == cache[i+1][j])
                answer.push_back(cache[i+1][j+1]);
        }
    }

    return answer;
}

File: JavaScript/level2/튜플.js
function sizeSort(a, b){
    return a.length - b.length;
}

function solution(s) {
    var answer = [];
    let values = []
    let temp = [];
    let val = 0;
    
    for(let i=2; i<s.length-1; i++){
        if(s[i] === '{')
            temp = [];
        else if(s[i] === '}'){
            temp.push(val);
            val = 0;
            values.push(temp);
        }
        else if(s[i] === ','){
            if(val != 0)
                temp.push(val);
            val = 0;
        }
        else{
            val *= 10;
            val += Number(s[i]);
        }
    }
    
    values.sort(sizeSort);
    answer.push(values[0][0]);
    
    for(let i=1; i<values.length; i++)
        answer.push(values[i].filter(v => !values[i-1].includes(v))[0]);
        
    return answer;
}